oracle/options.go

```diff
@@ -1,91 +1,98 @@
+// Specifically, this is what is utilized to construct price providers that are API based.
+func WithPriceAPIQueryHandlerFactory(factory types.PriceAPIQueryHandlerFactory) Option {
+	return func(m *OracleImpl) {
+		if factory == nil {
+			panic("api query handler factory cannot be nil")
 		}
 
-		o.logger = logger.With(zap.String("process", "oracle"))
+		m.priceAPIFactory = factory
 	}
 }
 
-// WithMetrics sets the metrics on the Oracle.
-func WithMetrics(metrics oraclemetrics.Metrics) Option {
-	return func(o *OracleImpl) {
-		if metrics == nil {
-			panic("cannot set nil metrics")
+// WithPriceWebSocketQueryHandlerFactory sets the websocket query handler factory for the oracle.
+// Specifically, this is what is utilized to construct price providers that are websocket based.
+func WithPriceWebSocketQueryHandlerFactory(factory types.PriceWebSocketQueryHandlerFactory) Option {
+	return func(m *OracleImpl) {
+		if factory == nil {
+			panic("websocket query handler factory cannot be nil")
 		}
 
-		o.metrics = metrics
+		m.priceWSFactory = factory
 	}
 }
 
-// WithMetricsConfig sets the metrics on the oracle from the given config.
-func WithMetricsConfig(config config.MetricsConfig) Option {
-	return func(o *OracleImpl) {
-		o.metrics = oraclemetrics.NewMetricsFromConfig(config)
+// WithMarketMapperFactory sets the market map factory for the oracle.
+// Specifically, this is what is utilized to construct market map providers.
+func WithMarketMapperFactory(factory mmclienttypes.MarketMapFactory) Option {
+	return func(m *OracleImpl) {
+		if factory == nil {
+			panic("market map factory cannot be nil")
+		}
+
+		m.marketMapperFactory = factory
 	}
 }
 
-// WithPriceAggregator sets the data aggregator on the Oracle.
-func WithPriceAggregator(agg PriceAggregator) Option {
-	return func(o *OracleImpl) {
-		if agg == nil {
-			panic("cannot set nil aggregator")
-		}
+// WithWriteTo sets the file path to which market map updates will be written to. Note that this is optional.
+func WithWriteTo(filePath string) Option {
+	return func(m *OracleImpl) {
+		m.writeTo = filePath
+	}
+}
 
-		o.priceAggregator = agg
+func WithPriceProviders(pps ...*types.PriceProvider) Option {
+	return func(m *OracleImpl) {
+		for _, pp := range pps {
+			m.priceProviders[pp.Name()] = ProviderState{Provider: pp}
+		}
 	}
 }
 
-// WithMarketMapGetter sets a getter function for the latest market map on the Oracle.
-func WithMarketMapGetter(fn func() mmtypes.MarketMap) Option {
-	return func(o *OracleImpl) {
-		o.marketMapGetter = fn
+func WithMetrics(met oraclemetrics.Metrics) Option {
+	return func(m *OracleImpl) {
+		m.metrics = met
 	}
 }
 
-// WithProviders sets the providers on the Oracle.
-func WithProviders(providers []*types.PriceProvider) Option {
-	return func(o *OracleImpl) {
-		o.providers = providers
+func WithMarketMapProvider(mmp *mmclienttypes.MarketMapProvider) Option {
+	return func(impl *OracleImpl) {
+		impl.mmProvider = mmp
 	}
```

Comment:
nit: can you add go docs

Reply:
Don't all oracle instances need to have a marketmap provider?



I think this should be a required argument rather than an `Option`

Reply:
actually, i forgot to remove this option. marketmap providers are instantiated via OracleConfig.Providers.

Reply:
added godocs to the other options [add godocs to other options](https://github.com/skip-mev/slinky/pull/503/commits/f9c7bd4f4371c2faa5e6a1bb2fc529736dcdded0)